models/src/main/scala/coop/rchain/models/EqualsM.scala

```diff
@@ -75,6 +75,17 @@ object EqualM extends EqualMDerivation {
 
   }
 
+  implicit def mapEqual[A: EqualM, B: EqualM]: EqualM[Map[A, B]] = new EqualM[Map[A, B]] {
+
+    override def equal[F[_]: Sync](self: Map[A, B], other: Map[A, B]): F[Boolean] = {
+      val pairsA = self.keys.toStream.zip(other.keys)
+      val pairsB = self.values.toStream.zip(other.values)
+      Sync[F].delay(self.size == other.size) &&^ pairsA
+        .forallM(tupled(EqualM[A].equal[F])) &&^ pairsB.forallM(tupled(EqualM[B].equal[F]))
```

Comment:
:heart_eyes:

Reply:
I'm probably nitpicking but use of `Sync[F].delay` is unnecesssary as it ties us to `Sync` for no good reason. `Applicative`  should be all that is needed here right?

```scala
((self.size == other.size).pure, pairsA.forallM(_), pairsB.forallM(_)).mapN(_ && _ && _)
```

or, if you want the optimization you're after `if (self.size == other.size) ... else false.pure` (Pelase note that this `&&^` is not even lazy in its arguments)

Reply:
We need `Sync` for `EqualM[A].equal[F]` being stack-safe as far as I know.

Reply:
Hmm, the only place where it might matter is typeclass derivation, right? But I am not sure why would it matter? Anyways, if this is the case then fine

Reply:
> Pelase note that this &&^ is not even lazy in its arguments

I think it's enough if the underlying F is delaying computations.
OTOH it's a good point, we should probably make `%%^` lazy in its second argument.

Reply:
Well - you can't know that :smile: More seriously, you still pay the price for constructing the "tree" But my question is: is `Sync` really needed here?

Reply:
Sync/Defer are a must if the computation can be costly, otherwise we'll incur the cost during computation construction. This is not the case in this particular EqualM instance, but can be an issue in others.

More importantly, we need Sync/Defer to delay the recursion, or else stackoverflow.

Technically, we could just require `Monad` and do 
```scala
def delayIfTrampolined(f: F[A]) = Applicative[F].unit.flatMap(_ => f)
```
but this would be a hack, and would only yield stacksafety for trampolined monads.

Reply:
Hmm - I am sorry, I still can't see it. Is there any place where you construct the computation for `equals` and don't want to use it right away? In cases like above you could've omitted the construction of some computations (as per the discussion above), yet this seemingly has never been a problem. As for delaying the recursion - 

```scala
import cats.{Applicative, Monad}

trait EqualM[F[_], A] {
  def eq(self: A, other: A): F[Boolean]
}

case class Rec(down: Option[Rec]) {
  override def toString: String = s"Rec(${down.fold("")(_ => "...")})"
}

object Rec {
  def nest(n: Int) = Rec {
    (0 to n).iterator.foldLeft[Option[Rec]](None)((down, _) => Some(Rec(down)))
  }
  implicit def eqRec[F[_]: Monad]: EqualM[F, Rec] = new EqualM[F, Rec] {
    override def eq(self: Rec, other: Rec): F[Boolean] =
      Monad[F].tailRecM((self.down, other.down)) {
        case (Some(r1), Some(r2)) => Monad[F].pure(Left((r1.down, r2.down)))
        case (r1, r2)             => Monad[F].pure(Right(r1 == r2))
      }
  }
}

val x1 = Rec.nest(100000)
val x2 = Rec.nest(100000)

import cats.instances.option._
Rec.eqRec[Option].eq(x1, x2)
```
yields

```
x1: Rec = Rec(...)
x2: Rec = Rec(...)

import cats.instances.`package`.option._
res0: Option[Boolean] = Some(true)
```

Reply:
> In cases like above you could've omitted the construction of some computations (as per the discussion above), yet this seemingly has never been a problem. 

As mentioned, `a.size == b.size` would not be a problem. I think we had a similar *actual* problem with `hugeList.hash` in HashM.

It's a good observation that recursion could be handled in the derivation using `tailrecM`. In fact, `forallM` is used anyway, so maybe it's just a matter of dropping the Sync constraint in the typeclass definition and seeing that the tests pass?

Still, EqualsM would only be stacksafe for stacksafe (trampolining) F-s, and this is mostly because tailrecM would cause a suspension. (BTW the tailrecM suspension is merely a convention not guarded by cats-laws, as exemplified by this [PR of mine to ZIO](https://github.com/zio/zio/pull/795/files)). Sync was used to make the `suspending monad` requirement explicit. 

https://github.com/rchain/rchain/blob/6dfde3f46ed67db35d303cde638784799d337aba/models/src/main/scala/coop/rchain/models/EqualsM.scala#L130-L138

Since delaying computations is needed at times, I'd lean to staying with sync and delaying all the computations rather than risking bugs due to inconsistency. Given the hack I mentioned above though, both approaches (Sync/tailrecM) can be made to work in all our use cases, and we have some tests for EqualsM stacksafety, hopefully. So we could change that.

Whatever you guys decide though, should most likely by implemented in a separate PR.

Reply:
> BTW the tailrecM suspension is merely a convention not guarded by cats-laws

Not necessarily:
https://github.com/typelevel/cats/blob/eb52c0bf90f81d60cd4120f3d10c62dd7dbbbd66/laws/src/main/scala/cats/laws/MonadLaws.scala#L39

But, I won't be obsessing over it. I think this requirement is "too much" (after all, `Sync` doesn't have much to say about stack-safety) but it's all in all harmless (minus slight performance penalty for `defer`\`delay`) and fixing it does not fit this PR. Thanks @ArturGajowy for the inspiring discussion

Reply:
Notice the test you linked to doesn't enforce stacksafety for running tailrecM with tailrecM - unlike the one I contributed to ZIO does:
https://github.com/zio/zio/pull/795/files#diff-b2ebf0e7fbc67d716ccbed5ecc008736R11-R17

I guess it doesn't fit the stacksafety requirements for tailrecM though, as no non-suspending monad can pass that test. Notice that very bug was caught by our stacksafety tests when I was taking ZIO for a ride in rholang tests.

> Sync doesn't have much to say about stack-safety

Stacksafety is achieved via suspension, which is the purpose of Sync.
Defer mentions suspending the computation too, but itself is not enough for our needs in the EqualsM/HashM (need Monad). Maybe we should just go with Monad + Defer?

Reply:
> Notice the test you linked to doesn't enforce stacksafety for running tailrecM with tailrecM

Ah, right. They sweep it under the carpet by saying: `constant stack space relative to f`.

> Stacksafety is achieved via suspension, which is the purpose of Sync.
Given that: `its implementation of flatMap is stack safe, meaning that you can describe tailRecM in terms of it ` then you can conclude that (it's the law I believe): `Sync.suspend(f) <-> Sync(f).flatten`. So, by the same reasoning, you can write everything you'd write with `suspend` using `flatMap` for any monad that's stack-safe (or maybe I am missing something here)

But, ok, I'm buying your arguments - with "normal" `Monad.tailRecM` you have to code carefully. I do not have objections then